launch/launch/launch_service.py

```diff
@@ -0,0 +1,277 @@
+                self._process_one_event())
+            if self.__shutting_down:
+                # If shutting down and idle then we're done.
+                if is_idle:
+                    process_one_event_task.cancel()
+                    return
+                else:
+                    entity_futures = [pair[1] for pair in self._entity_future_pairs]
+                    entity_futures.append(process_one_event_task)
+                    entity_futures.extend(self.__context._completion_futures)
+                    done: Set[asyncio.Future] = set()
+                    while not done:
+                        done, pending = await asyncio.wait(
+                            entity_futures,
+                            loop=self.__loop_from_run_thread,
+                            timeout=1.0,
+                            return_when=asyncio.FIRST_COMPLETED)
+                        if not done:
+                            _logger.debug('still waiting on futures: {}'.format(entity_futures))
+            else:
+                await process_one_event_task
+
+    def run(self, *, shutdown_when_idle=True) -> None:
+        """
+        Start the event loop and visit all entities of all included LaunchDescriptions.
+
+        This should only ever be run from a single thread.
+
+        :param: shutdown_when_idle if True (default), the service will shutdown when idle
+        """
+        # Make sure this has not been called in multiple threads.
+        with self.__running_lock:
+            if self.__running:
+                raise RuntimeError(
+                    'LaunchService.run() called from multiple threads concurrently.')
+            self.__running = True
+
+        self.__shutdown_when_idle = shutdown_when_idle
+
+        # Acquire the lock and initialize the asyncio loop.
+        with self.__loop_from_run_thread_lock:
+            self.__loop_from_run_thread = osrf_pycommon.process_utils.get_loop()
+            if self.__loop_from_run_thread is None:
+                raise RuntimeError('__loop_from_run_thread unexpectedly None')
+            if self.__debug:
+                self.__loop_from_run_thread.set_debug(True)
+            self.__context._set_asyncio_loop(self.__loop_from_run_thread)
+
+        # Run the asyncio loop over the main coroutine that processes events.
+        try:
+            sigint_received = False
+            run_loop_task = self.__loop_from_run_thread.create_task(self.__run_loop())
+            while not run_loop_task.done():
+                try:
+                    self.__loop_from_run_thread.run_until_complete(run_loop_task)
+                except KeyboardInterrupt:
+                    base_msg = 'user interrupted with ctrl-c (SIGINT)'
+                    if not sigint_received:
+                        _logger.warn(base_msg)
+                        self._shutdown(reason='ctrl-c (SIGINT)', due_to_sigint=True)
+                        sigint_received = True
+                    else:
```

Comment:
This is the opposite to what we've done in the previous `launch`: https://github.com/ros2/launch/pull/58/files
I think you've mentioned before ROS 1 does a similar thing, escalating based on the number of interrupts received, is that right?

Reply:
The signal handling of the service needs more work, I also don't handle SIGTERM, which I probably should do in order to avoid zombie processes.

I'll spend some time on this right now, but at some point I'll need to just start making issues for things, otherwise this pr will never get merged in time.

Reply:
To me it's fine if this is one of the things that gets ticketed (it will be easier to review as an incremental improvement anyway). My intent is to point out what I see while reviewing (because otherwise we might forget), then we can make the judgement on them individually.

Reply:
For sure, didn't mean to sound like I was complaining, just saying I might fix it up some now and ticket other stuff for later, even if it ends up in the release as a follow and smaller change.

Reply:
I pushed some enhancements in 010edcb, one of which is that repeated ctrl-c does not exit early.